### PR TITLE
Adding android authentication post

### PR DESCRIPTION
**Pull request by bdemers**
This blog post has the following features:

- [x] A GitHub Repository with a polished README
- [ ] A title that's approved by Dev Marketing
- [ ] A URL that's approved by Dev Marketing
- [x] Rendered locally and confirmed that no Markdown typos exist

### CODE REVIEW

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+- Our security experts take the development effort of creating a secure authentication method off your plate
+- OAuth 2.0 and OpenID Connect allows your users to easily authenticate on your app with social login, using industry-standard secure protocols
+
+This post will show you a simple example of how you can use Okta OIDC SDK to create a (Kotlin) Android app with authentication.
+
+toc::[]
+
+== (TLDR) Download the finished app with Android Authentication
+
+If you want to follow along using a completed version of the app,clone this repository from  https://github.com/oktadeveloper/okta-oidc-android-example[GitHub] using the `simplified` branch:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:32]:
```suggestion
If you want to follow along using a completed version of the app, clone this repository from https://github.com/oktadeveloper/okta-oidc-android-example[GitHub] using the `simplified` branch:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+
+[source,sh]
+----
+git clone https://github.com/oktadeveloper/okta-oidc-android-example.git -b simplified
+cd okta-oidc-android-example
+----
+
+NOTE: The finished app in the `master` branch from the repository includes everything described in this blog, along with  some additional UI and functionality, testing Okta OIDC easier, and providing you with a production-level skeleton to work with when creating your own app.
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:40]:
```suggestion
NOTE: The finished app in the `master` branch from the repository includes everything described in this blog, along with some additional UI and functionality, testing Okta OIDC easier, and providing you with a production-level skeleton to work with when creating your own app.
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+
+== Creating your first Android app
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:42]: http://www.conissaunce.com/pages/infoq-title-case.html says this should be `Creating Your First Android App`. If I channel Lindsay, it should never have `ing`, so should be `Create Your First Android App`.

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+
+First, you'll need to download and install https://developer.android.com/studio[Android Studio].
+Next, launch the app and navigate  to **File** → **New...** → **New Project...**. Then, create an "Empty Activity" for "Phone and Tablet". You should now see a screen similar to this:
+
+image::{% asset_path 'blog/android-authentication/android-studio-new-project.png' %}[alt=Create new Android Project,width=700,align=center]
+
+Choose a "Package name" related to a domain you own (I own akaita.com, therefore I used `com.akaita.myapplication` in the example), a **Save location** you can remember, "Kotlin", and a "Minimum API level" of `23`. Finally, click **Finish**.
+
+[#okta-oidc-config]
+== Get your Okta OIDC configuration settings
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:52]: Needs more capitalization

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+
+Before adding anything else to your app, you'll need to quickly obtain the configuration parameters required to configure Okta OIDC authentication:
+
+1. Log in to your Okta Developer account or https://developer.okta.com/[sign up for a **free** developer account]
+2. Once you've logged on, you'll see the Admin Dashboard. Take a note of your "Org URL" (at the top right of the screen). In my case, I see: `\https://dev-123456.okta.com`
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:57]:
```suggestion
2. Once you've logged on, you'll see the Admin Dashboard. Take note of your "Org URL" (at the top right of the screen). In my case, I see: `https://dev-123456.okta.com`
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+3. Navigate to **Applications** → **Add Application**
+4. Choose **Native** as the platform
+5. Add a **Login redirect URIs** similar to `com.okta.dev-123456:/callback` (using your own `dev-XXXX`)
+6. Add a **Logout redirect URIs** similar to `com.okta.dev-123456:/logout` (using your own `dev-XXXX`)
+7. Click on **Done**
+
+image::{% asset_path 'blog/android-authentication/okta-create-new-application.jpg' %}[alt=Create a new Okta application,width=700,align=center]
+
+Next, take note of the **Client ID** that's found in the **General** tab of your new Okta application entry. It's a long string of around 20 characters.
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:66]:
```suggestion
Next, take note of the **Client ID** that's found in the **General** tab of your new Okta application. It's a long string of around 20 characters.
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+== Add a sign-in button to your Android app
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:68]: Needs capitalization

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+        android:text="Sign in" />
+
+    </LinearLayout>
+
+</ScrollView>
+----
+
+This button will be used to authenticate using OAuth 2.0 + OpenID Connect, thanks to the Okta OIDC SDK.
+
+== Give your Android app permission to use the Internet
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:101]: Capitalization

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+
+In your `app/src/main/AndroidManifest.xml` add between the `<manifest></manifest>` tags:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:103]:
```suggestion
In your `app/src/main/AndroidManifest.xml` add the following XML between the `<manifest></manifest>` tags:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+
+[source,xml]
+----
+<uses-permission android:name="android.permission.INTERNET" />
+----
+
+== Add Okta OIDC SDK to your Android app
+
+1. Add a `manifestPlaceholder` for `appAuthRedirectScheme` in `app/build.gradle`. Make sure it is consistent with your Redirect URIs. For instance, my Redirect URIs look like `com.okta.dev-123456:/callback`, therefore my `appAuthRedirectScheme` is `com.okta.dev-123456`
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:112]: I don't think "Redirect" should be capitalized in this sentence.

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+
+2. Add the required Okta dependencies in `app/build.gradle`, and declare a Java 1.8 target in the `compileOptions` clause (this is required by the OIDC library)
+
+====
+[source,kotlin]
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:117]: This looks like Groovy, not Kotlin. If it was Kotlin, wouldn't it have a `.kt` suffix?

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+       release {
+           minifyEnabled false
+           proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+       }
+   }
+}
+
+dependencies {
+   implementation fileTree(dir: 'libs', include: ['*.jar'])
+   implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:155]: Indentation is inconsistent. Above, there's 4 and 8 spaces, here there's 3.

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+}
+----
+
+<1> Keep you `applicationId` here.
+<2> The redirect URI for the application you created in your Okta Developer Console.
+<3> Okta OIDC libraries require Java 1.8 compatibility.
+<4> Add the dependencies required for the Okta OIDC library.
+====
+
+== Configure Okta OIDC in your app
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:182]: I'd add Android (for SEO) and capitalization

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+In your `MainActivity` class, add a couple of class properties anda couple of new methods,calling them from `onCreate`:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:183]:
```suggestion
In your `MainActivity` class, add a couple of class properties and a couple of new methods, calling them from `onCreate()`:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+private lateinit var webAuth: WebAuthClient // <1>
+
+/**
+* The authorized client to interact with Okta's endpoints.
+*/
+private lateinit var sessionClient: SessionClient // <2>
+
+override fun onCreate(savedInstanceState: Bundle?) {
+   super.onCreate(savedInstanceState)
+   setContentView(R.layout.activity_main)
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:213]: Has three spaces, should be 4.

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 16]
```diff
@@ -0,0 +1,508 @@
...
+<4> `private fun setupWebAuthCallback()` is the place where you can define the action to take when authentication succeeds, fails or is canceled...
+====
+
+TIP: You can create a https://github.com/oktadeveloper/okta-oidc-android-example/blob/e34375dccd9a6898ecf62ecc4c63d08e63957a57/app/src/main/java/com/okta/oidc/example/Extensions.kt#L24[utility method] to automatically detect if your application is running in an emulator.
+
+Make sure to use the values you wrote down during the link:#okta-oidc-config[Get your Okta OIDC configuration settings] step to create `val oidcConfig` in `private fun setupWebAuth()`.
+
+Now that you have a minimal configuration of Okta OIDC, it's time to add a listener for the button we added previously.
+
+== Log in using Okta OIDC
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:277]: Capitalization using http://www.conissaunce.com/pages/infoq-title-case.html

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+
+Create a `AuthenticationPayload` and call `webAuth.signIn` in `MainActivity.kt`.
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:279]:
```suggestion
Create a payload using `AuthenticationPayload` and call `webAuth.signIn()` in `MainActivity.kt`.
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 18]
```diff
@@ -0,0 +1,508 @@
...
+When you launch your app and click on the `SIGN IN` button, you will be greeted with OKta's authentication portal, which you can use to authenticate users into your app:
+
+image::{% asset_path 'blog/android-authentication/android-browser-login.png' %}[alt=Android browser login prompt,width=400,align=center]
+
+== (Optional) Adding Social login
+
+It should be easy for users to authenticate into your app. Okta OIDC SDK helps you accomplish this by allowing users to use their social accounts to validate their identity..
+
+With Okta, you can add an *external Identity Provider*—such as Google, Facebook, LinkedIn or Microsoft.
+To achieve this, connect to the external identity providers and ask them to trust Okta for your application.   This is done in three simple steps:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:341]:
```suggestion
To achieve this, connect to the external identity providers and ask them to trust Okta for your application. This is done in three simple steps:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 19]
```diff
@@ -0,0 +1,508 @@
...
+signIn.setOnClickListener {
+   val payload = AuthenticationPayload.Builder()
+       .setIdp("{IdP-id}") // From your "Okta" admin console
+       .setIdpScope("clientScope1", "clientScope2", "clientScope3") // Optional, in case the IDP requires it
+       .build()
+   webAuth.signIn(this, payload)
+}
+----
+
+Now, when a user clicks your **SIGN IN** button, they'll be greeted with the login page of the IDP of your choice, which will be used to authenticate users into your app.
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:360]: `SIGN IN` is used above - be consistent

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 20]
```diff
@@ -0,0 +1,508 @@
...
+
+== (Optional) Biometric login
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:362]: Capitalize login

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 21]
```diff
@@ -0,0 +1,508 @@
...
+
+Additionally, biometrics can be used to access sessions  created by Okta OIDC.
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:364]:
```suggestion
Additionally, biometrics can be used to access sessions created by Okta OIDC.
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+The `BiometricPrompt` recently published by the Android team makes it a very feasible option, taking a lot of the complexity off your hands and offering a unified familiar & native experience for users on all variants of Android (Google devices, Samsung devices, ...).
+
+NOTE: `BiometricPrompt` uses the available resources in each device to offer whichever options are possible. Those include Iris authentication, fingerprint authentication, PIN authentication, Pattern authentication, etc...
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:367]:
```suggestion
NOTE: `BiometricPrompt` uses the available resources in each device to offer whichever options are possible. Those include Iris authentication, fingerprint authentication, PIN authentication, Pattern authentication, etc.
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 23]
```diff
@@ -0,0 +1,508 @@
...
+
+In essence, there are two components to be taken  into account:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:369]:
```suggestion
In essence, there are two components to be taken into account:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 24]
```diff
@@ -0,0 +1,508 @@
...
+       .withConfig(oidcConfig)
+       .withContext(applicationContext)
+       .withCallbackExecutor(null)
+       .withEncryptionManager(keyguardEncryptionManager)
+       .create()
+   sessionClient = webAuth.sessionClient
+}
+----
+
+Now, whenever a user opens a new instance of your app, they will be able to access it using their biometric information. In the below example, a Google Pixel phone, the authentication is a  fingerprint:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:481]:
```suggestion
Now, whenever a user opens a new instance of your app, they will be able to access it using their biometric information. In the below example, a Google Pixel phone, the authentication is a fingerprint:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 25]
```diff
@@ -0,0 +1,508 @@
...
+
+image::{% asset_path 'blog/android-authentication/android-fingerprint-prompt.png' %}[alt=Android fingerprint prompt,width=400,align=center]
+
+== Summary
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:485]: Can this be made better for SEO? Most of the time, we combine the summary and the learn more.

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 26]
```diff
@@ -0,0 +1,508 @@
...
+
+It's now easier than ever to implement OAuth 2.0 and OpenID authorization thanks to Okta OIDC SDK. A high effort, high maintenance chore just became a very straightforward task.
+
+By taking advantage of `BiometricPrompt` and Kotlin, we can make our apps even safer, while still creating a seamless user experience that  feels native to each manufacturer's UI style.
+
+Although the example we've created here does enough to satisfy the needs of the majority of apps, Okta OIDC SDK doesn't stop there. The https://github.com/okta/okta-oidc-android[Okta OIDC Android repository] contains  a variety of ideas and suggestions to improve user experience like:
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:491]:
```suggestion
Although the example we've created here does enough to satisfy the needs of the majority of apps, Okta OIDC SDK doesn't stop there. The https://github.com/okta/okta-oidc-android[Okta OIDC Android repository] contains a variety of ideas and suggestions to improve user experience like:
```

File: _source/_posts/2020-04-20-android-authentication.adoc [hunk 27]
```diff
@@ -0,0 +1,508 @@
...
+
+- Using your own OkHttp client
+- Using a custom UI to log in
+- Having fine-grained control over session tokens' expiration, refresh, etc...
+- Settings to handle preference of browser client for the authentication process (Chrome, Samsung browser, Firefox, etc...)
+
+== Further learning
```
Comment (mraible) [inline _source/_posts/2020-04-20-android-authentication.adoc:498]: This is pretty generic. I'd change it to be something like "Learn More About Android Auth and OIDC"

File: _source/_posts/2020-04-20-android-authentication.adoc (hunk 5 above)
Comment (bdemers) [reply _source/_posts/2020-04-20-android-authentication.adoc:57]: The backslack prevents the automatic href creation

Comment (bdemers) [reply _source/_posts/2020-04-20-android-authentication.adoc:57]: (i'll remove the extra `a` though)

File: _source/_posts/2020-04-20-android-authentication.adoc (hunk 19 above)
Comment (bdemers) [reply _source/_posts/2020-04-20-android-authentication.adoc:360]: Changed other ref to be **SIGN IN**

**Pull request merged**